weedcoco/index/indexing.py

```diff
@@ -0,0 +1,149 @@
+            for field in variable_to_null_fields:
+                if agcontext.get(field) == "variable":
+                    del agcontext[field]
+
+        for annotation in coco["annotations"]:
+            image = id_lookup["images", annotation["image_id"]]
+            image.setdefault("annotations", []).append(annotation)
+            annotation["category"] = id_lookup["categories", annotation["category_id"]]
+            # todo: add collection, license
+            _flatten(annotation["category"], annotation, "category")
+            # todo: add collection from collection_memberships
+            image["thumbnail"] = str(
+                self.thumbnail_dir
+                / os.path.basename(image["file_name"])[:2]
+                / os.path.basename(image["file_name"])
+            )
+
+        for image in coco["images"]:
+            try:
+                image["resolution"] = image["width"] * image["height"]
+            except KeyError:
+                pass
+            image["agcontext"] = id_lookup["agcontexts", image["agcontext_id"]]
+            image["sortKey"] = hash(
+                image["file_name"]
+            )  # for deterministic random order
+            _flatten(image["agcontext"], image, "agcontext")
+            # todo: add license
+            image["task_type"] = set()
+            for annotation in image["annotations"]:
+                for k in annotation:
+                    image.setdefault(f"annotation__{k}", []).append(annotation[k])
+
+                # determine available task types
+                image["task_type"].add("classification")
+                if "segmentation" in annotation:
+                    image["task_type"].add("segmentation")
+                    image["task_type"].add("bounding box")
+                if "bbox" in annotation:
+                    image["task_type"].add("bounding box")
+            image["task_type"] = sorted(image["task_type"])
+
+        self.indexes = coco
+
+    def generate_batches(self):
+        """
+        Split indexes into batches to reduce payload size of each request to ElasticSearch
+        """
+        indexes = self.indexes
+        if "images" in indexes and len(indexes["images"]) > 0:
+            images = indexes["images"]
+            for start in range(0, len(images), self.batch_size):
+                blobs = []
+                batch = images[start : start + self.batch_size]
+                for image in batch:
+                    blobs.append(
+                        {
+                            "_index": self.es_index_name,
+                            "_type": self.es_type_name,
+                            "_source": image,
+                        }
+                    )
+                yield blobs
+
+    def post_to_index(self):
+        """
+        Send post request to ElasticSearch
+        """
+        for index_batch in self.generate_batches():
+            helpers.bulk(self.es_client, index_batch)
+
```

Comment:
I'm not sure what exceptions need to be logged here.
Possible results are as https://elasticsearch-py.readthedocs.io/en/7.9.1/exceptions.html